File: src/go/problems/reverse_vowels_test.go
// https://leetcode.com/problems/reverse-vowels-of-a-string/

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type ReverseVowels struct{}

func TestReverseVowels(t *testing.T) {
	gen := core.TestSuite[ReverseVowels]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("hello").Result("holle")
	}).Add(func(tc *core.TestCase) {
		tc.Param("leetcode").Result("leotcede")
	}).Add(func(tc *core.TestCase) {
		tc.Param(" ").Result(" ")
	}).Add(func(tc *core.TestCase) {
		tc.Param("appear").Result("appear")
	}).Run(t)
}

func (ReverseVowels) Solution(s string) string {
	var vowels = map[byte]struct{}{'a': {}, 'e': {}, 'i': {}, 'o': {}, 'u': {}, 'A': {}, 'E': {}, 'I': {}, 'O': {}, 'U': {} }

	result := []byte(s)
	left := 0
	right := len(s) - 1
	for left < right {
		if _, ok := vowels[s[left]]; ok {
			if _, ok := vowels[s[right]]; ok {
				result[left], result[right] = result[right], result[left]
				left++
			}

			right--
		} else {
			left++
		}
	}

	return string(result)
}

File: src/python/problems/reverse_vowels_test.py
# https://leetcode.com/problems/reverse-vowels-of-a-string

from core.problem_base import *

class ReverseVowels(ProblemBase):
    def Solution(self, s: str) -> str:
        map = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}

        result = list(s)
        left = 0
        right = len(s) - 1
        while left < right:
            if s[left] in map:
                if s[right] in map:
                    result[left], result[right] = result[right], result[left]
                    left += 1
                right -= 1
            else:
                left += 1
        
        return ''.join(result)

if __name__ == '__main__':
    TestGen(ReverseVowels) \
        .Add(lambda tc: tc.Param("s", "hello").Result("holle")) \
        .Add(lambda tc: tc.Param("s", "leetcode").Result("leotcede")) \
        .Add(lambda tc: tc.Param("s", " ").Result(" ")) \
        .Run()

